### ISSUE TITLE
Get Request Precondition Support

### ISSUE THREAD

**Issue by tustvold** [apache/arrow-rs#2241]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

There are 4 broadly supported request preconditions:

* If-Match
* If-Not-Match
* If-Not-Modified-Since
* If-Unmodified-Since

https://cloud.google.com/storage/docs/request-preconditions
https://docs.microsoft.com/en-us/rest/api/storageservices/specifying-conditional-headers-for-blob-service-operations
https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObject.html#API_GetObject_RequestSyntax

**Describe the solution you'd like**

I would like to propose adding

```
#[derive(Debug, Default)]
struct GetOptions {
    pub if_match: Option<String>,
    pub if_not_match: Option<String>,
    pub if_not_modified_since: Option<DateTime<Utc>>,
    pub if_unmodified_since: Option<DateTime<Utc>>,
}

pub trait ObjectStore {
     async fn get_opts(&self, location: &Path, options: GetOptions) -> Result<GetResult>

     async fn get_range_opts(&self, location: &Path, range: Option<usize>, options: GetOptions) -> Result<Bytes>
     ...
}

pub enum Error {
    #[snafu(display("Precondition failed: {}", source))]
    Precondition {
        source: Box<dyn std::error::Error + Send + Sync + 'static>,
    }

    #[snafu(display("Not modified: {}", source))]
    NotModified {
        source: Box<dyn std::error::Error + Send + Sync + 'static>,
    }
}
```

**Describe alternatives you've considered**

We could not add support for this

**Additional context**

#2230

**Comment by crepererum** [comment]
How do you model stores that don't support these checks? Do they just return "unsupported" errors?

**Comment by tustvold** [comment]
All stores support these for get requests.

**Comment by crepererum** [comment]
> All stores support these for get requests.

I think this crate also provides some extendable interface, i.e. users can implement their own  stores and use for example `etcd` as a storage backend. But I suppose every store could implement this check client-side if the "other end" doesn't support it, so this is OK then.

**Comment by tustvold** [comment]
Yeah, I would expect such stores to return `Error::NotSupported` if the feature cannot be supported, and `Error::NotImplemented` if it simply has not been implemented

**Comment by alamb** [comment]
Migrating from arrow-rs issue [#2241](https://github.com/apache/arrow-rs/issues/2241)
